src/python/pants/backend/terraform/dependencies.py

```diff
@@ -0,0 +1,173 @@
+from pants.engine.process import FallibleProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import (
+    DependenciesRequest,
+    SourcesField,
+    Targets,
+    WrappedTarget,
+    WrappedTargetRequest,
+)
+
+
+@dataclass(frozen=True)
+class TerraformDependenciesRequest:
+    source_files: SourceFiles
+    directories: Tuple[str, ...]
+    backend_config: SourceFiles
+    dependencies_files: SourceFiles
+
+    # Not initialising the backend means we won't access remote state. Useful for `validate`
+    initialise_backend: bool = False
+
+
+@dataclass(frozen=True)
+class TerraformDependenciesResponse:
+    fetched_deps: Tuple[Tuple[str, Digest], ...]
+
+
+@rule
+async def get_terraform_providers(
+    req: TerraformDependenciesRequest,
+) -> TerraformDependenciesResponse:
+    args = ["init"]
+    if req.backend_config.files:
+        args.append(
+            terraform_arg(
+                "-backend-config",
+                terraform_relpath(req.directories[0], req.backend_config.files[0]),
+            )
+        )
+        backend_digest = req.backend_config.snapshot.digest
+    else:
+        backend_digest = EMPTY_DIGEST
+
+    args.append(terraform_arg("-backend", str(req.initialise_backend)))
+
+    with_backend_config = await Get(
+        Digest,
+        MergeDigests(
+            [
+                req.source_files.snapshot.digest,
+                backend_digest,
+                req.dependencies_files.snapshot.digest,
+            ]
+        ),
+    )
+
+    # TODO: Does this need to be a MultiGet? I think we will now always get one directory
+    fetched_deps = await MultiGet(
+        Get(
+            FallibleProcessResult,
+            TerraformProcess(
+                args=tuple(args),
+                input_digest=with_backend_config,
+                output_files=(".terraform.lock.hcl",),
+                output_directories=(".terraform",),
+                description="Run `terraform init` to fetch dependencies",
+                chdir=directory,
+            ),
+        )
+        for directory in req.directories
+    )
+
+    return TerraformDependenciesResponse(
+        tuple(zip(req.directories, (x.output_digest for x in fetched_deps)))
+    )
+
+
+@dataclass(frozen=True)
+class TerraformInitRequest:
+    root_module: TerraformRootModuleField
+    backend_config: TerraformBackendConfigField
+    dependencies: TerraformDependenciesField
+
+    # Not initialising the backend means we won't access remote state. Useful for `validate`
+    initialise_backend: bool = False
+
+
+@dataclass(frozen=True)
+class TerraformInitResponse:
+    sources_and_deps: Digest
+    terraform_files: tuple[str, ...]
+    chdir: str
+
+
+@rule
+async def init_terraform(request: TerraformInitRequest) -> TerraformInitResponse:
+    address_input = request.root_module.to_address_input()
+    module_address = await Get(Address, AddressInput, address_input)
```

Comment:
I would recommend inferring `root_module` to point at the `terraform_module` in the same directory if it is not specified. This is what `go_binary` does for the `main` field. (There should only be one `terraform_module` per directory so this inference should always work.)

Reply:
But this is fine to be done as a follow-up.